test/infrastructure/docker/controllers/dockermachine_controller.go

```diff
@@ -198,6 +202,32 @@ func (r *DockerMachineReconciler) reconcileNormal(machine *clusterv1.Machine, do
 	providerID := externalMachine.ProviderID()
 	dockerMachine.Spec.ProviderID = &providerID
 
+	// Set the nodeRef on the machine
+	nodeList := &v1.NodeList{}
+	if err := r.Client.List(ctx, nodeList); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to list noderef")
+	}
+	pid, err := noderefutil.NewProviderID(providerID)
+	if err != nil {
+		r.Log.Error(err, "failed to parse node's provider spec")
+		return ctrl.Result{}, err
+	}
+	for _, node := range nodeList.Items {
+		npid, err := noderefutil.NewProviderID(node.Spec.ProviderID)
+		if err != nil {
+			r.Log.Error(err, "failed to parse node's provider spec")
+			continue
+		}
+		if pid.Equals(npid) {
+			machine.Status.NodeRef = &v1.ObjectReference{
```

Comment:
The Machine controller should be setting the `NodeRef` based on matching the `providerID` of the node with the `providerID` of the `Machine`/`DockerMachine`.

This would normally be done through the cloud provider integration. I'm wondering if CAPD should have a `Node` controller that sets the `providerID` on Nodes to provide this functionality.

Reply:
we do set the provider id, but i noticed we don't set the node ref on the DockerMachine. I was mostly copying AWS here. The AWSMachine controller sets the node ref on the AWSMachine and then it gets copied over into the machine via the machine controller. I don't think we need a node controller or a cloud-provider (yet) but i think it might be nice if someone wanted to learn about implementing a cloud provider.

Reply:
oops, hold on, what i said doesn't look right anymore. let me see what i was copying...

Reply:
ok i need to debug something, i got confused and the machine controller definitely sets the noderef, sorry. let me debug why it's not getting set.

Reply:
ok i see what you're saying. the *nodes* don't have the right provider ID